### PR TITLE
Fix Windows Tests

### PR DESCRIPTION
**Pull request by varunagrawal**
My attempt at fixing #1541.

1. Add fix for compilation in Visual Studio. This seems to be a weird issue with how `size_t` is interpreted since the `multiple symbols defined` error comes from additional declarations of `std::map<Key, size_t>`. This might be a bug in MSVC 16.
2. Fixed some more tests across the repo.
3. Updated the Windows CI to run additional GTSAM tests as well as some `gtsam_unstable` tests.

There is still quite a bit of work to do to get all the tests passing on Windows, but hopefully this is a good start point.

### CODE REVIEW

**Comment by talregev** [comment]
@varunagrawal You asked my help for these tests.
I can help you to compile them, but not run them (make the tests correct). 
I found some bug that I can share with you:
1. On this line: https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/CMakeLists.txt#L111
It should be `GTSAM_EXPORTS` instead of `GTSAM_UNSTABLE_EXPORTS`. There isn't definition for `GTSAM_UNSTABLE_EXPORTS`. I think gtsam_unstable should import/export symbol as gtsam lib.
When I switch it, it make the dll import when compile gtsam_unstable. So the bug is deeper.
2. On this line: https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/CMakeLists.txt#L107
It should be `GTSAM_IMPORT_STATIC` instead GTSAM_UNSTABLE_IMPORT_STATIC. I didn't try it.
I think it also be confuse, and need to investigate more deeply the bug. 
3. On this line https://github.com/borglab/gtsam/blob/develop/gtsam/hybrid/tests/testHybridGaussianFactorGraph.cpp#L46
It should add `#include <numeric>` for handle the compile error: `error C2039: 'iota': is not a member of 'std'`

Good luck to hunt more bugs!

**Comment by talregev** [comment]
By the way, I am not sure if you notice. Even that the ci pass, some of the tests are not compile with missing symbols.

**Comment by talregev** [comment]
testBasisFactors.cpp.obj : error LNK2019: unresolved external symbol "class Eigen::Matrix<double,-1,-1,0,-1,-1> __cdecl gtsam::kroneckerProductIdentity(unsigned __int64,class Eigen::Matrix<double,1,-1,1,1,-1> const &)" (?kroneckerProductIdentity@gtsam@@YA?AV?$Matrix@N$0?0$0?0$0A@$0?0$0?0@Eigen@@_KAEBV?$Matrix@N$00$0?0$00$00$0?0@3@@Z) referenced in function "protected: void __cdecl gtsam::Basis<class gtsam::Chebyshev2>::VectorDerivativeFunctor::calculateJacobian(void)" (?calculateJacobian@VectorDerivativeFunctor@?$Basis@VChebyshev2@gtsam@@@gtsam@@IEAAXXZ)

**Comment by varunagrawal** [comment]
> 1. On this line: [`develop`/gtsam_unstable/CMakeLists.txt#L111](https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/CMakeLists.txt?rgh-link-date=2023-06-29T16%3A08%3A32Z#L111) It should be `GTSAM_EXPORTS` instead of `GTSAM_UNSTABLE_EXPORTS`. There isn't definition for `GTSAM_UNSTABLE_EXPORTS`. I think gtsam_unstable should import/export symbol as gtsam lib. When I switch it, it make the dll import at compilation of gtsam_unstable. Need to investigate more deeply the bug.

`GTSAM_EXPORTS` and `GTSAM_UNSTABLE_EXPORTS` are autogenerated as part of `dllexport.h.in`. So this is not a bug.

> 2. On this line: [`develop`/gtsam_unstable/CMakeLists.txt#L107](https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/CMakeLists.txt?rgh-link-date=2023-06-29T16%3A08%3A32Z#L107) It should be `GTSAM_IMPORT_STATIC` instead GTSAM_UNSTABLE_IMPORT_STATIC. I didn't try it. I think it also do import instead export and need to investigate more deeply the bug.

Same reasoning as above. :)

> 3. On this line [`develop`/gtsam/hybrid/tests/testHybridGaussianFactorGraph.cpp#L46](https://github.com/borglab/gtsam/blob/develop/gtsam/hybrid/tests/testHybridGaussianFactorGraph.cpp?rgh-link-date=2023-06-29T16%3A08%3A32Z#L46) It should add `#include <numeric>` for handle the compile error: `error C2039: 'iota': is not a member of 'std'`

Nice catch!

**Comment by varunagrawal** [comment]
> By the way, I am not sure if you notice. Even that the ci pass, some of the tests are not compile with missing symbols.

Yup I am aware. That's the reason this PR is still a draf. :)

**Comment by talregev** [comment]
> `GTSAM_EXPORTS` and `GTSAM_UNSTABLE_EXPORTS` are autogenerated as part of `dllexport.h.in`. So this is not a bug.
> 
> >

@varunagrawal Thank you the explanation. I was doing second review to check, and I understand it my own that it ok. I edit my comment, and leave the original msg with strikes.

**Comment by talregev** [comment]
@varunagrawal Are you going to continue to work on these? Is it low priority task?

File: gtsam/inference/Ordering.h
Comment (talregev) [inline gtsam/inference/Ordering.h:74]: should remove.

File: gtsam/inference/Ordering.h
Comment (talregev) [inline gtsam/inference/Ordering.h:78]: should remove

**Review by talregev — CHANGES_REQUESTED** [review]

File: gtsam/inference/Ordering.h
Comment (talregev) [inline gtsam/inference/Ordering.h:73]: `GTSAM_EXPORT` should removed.

File: gtsam/inference/Ordering.h
Comment (talregev) [inline gtsam/inference/Ordering.h:77]: `GTSAM_EXPORT` should removed.

**Review by talregev — CHANGES_REQUESTED** [review]

File: gtsam/discrete/SignatureParser.h
Comment (talregev) [inline gtsam/discrete/SignatureParser.h:24]: `#include <gtsam/dllexport.h>` should removed. There exactly one there.

**Comment by talregev** [comment]
After my fix test PR was merge. What fix this PR is aiming for?

**Comment by varunagrawal** [comment]
A lot of tests in the CI are commented out. We should fix all of them in order to merge this PR.

**Comment by talregev** [comment]
> A lot of tests in the CI are commented out. We should fix all of them in order to merge this PR.

It should focus on one test fix / one type of problem. Is a bad practice to create a PR that deal with all the problems.
Also what current changes are aim to fix? What problem they are solving?

**Comment by varunagrawal** [comment]
@talregev @mikesheffler @dellaert I fixed all the `GTSAM` tests for Windows. Only the tests for `GTSAM_UNSTABLE` are left which I will leave for later.

**Comment by varunagrawal** [comment]
> You should squash the commits. It not recommended to have all these commits and merges to dev branch.

We prefer to keep the history so we can go back and identify bad commits and undo them. This is why we don't squash or force-push.

**Comment by varunagrawal** [comment]
> @varunagrawal Are you interesting to deal with these errors? image

Why is it using `gtsamDebug.dll` in the Release config?

**Comment by talregev** [comment]
> > @varunagrawal Are you interesting to deal with these errors? image
> 
> Why is it using `gtsamDebug.dll` in the Release config?

This is another bug you have in the cmake when I compile it with vcpkg.
I am not sure what happen in the regular compilation. 

But the debug assertion still happen on your side. Are you interesting to solve it?

**Comment by varunagrawal** [comment]
> This is another bug you have in the cmake when I compile it with vcpkg. I am not sure what happen in the regular compilation.

I have never had this issue. I have been compiling and debugging using the Release config for the last week and it seems to be fine. Maybe the issue is in your vcpkg settings?

> But the debug assertion still happen on your side. Are you interesting to solve it?

I do not get this error on my end. I re-ran the tests again and they pass without problems.

**Comment by talregev** [comment]
> I have never had this issue. I have been compiling and debugging using the Release config for the last week and it seems to be fine. Maybe the issue is in your vcpkg settings?

Let's deal with this later. It not for this PR. 

> I do not get this error on my end. I re-ran the tests again and they pass without problems.

Of course there isn't errors. It not enable. I can enable it and you will see these errors.

**Comment by varunagrawal** [comment]
> Of course there isn't errors. It not enable. I can enable it and you will see these errors.
Sorry I don't understand. What is not enabled?

**Comment by talregev** [comment]
> Sorry I don't understand. What is not enabled?

The flag that enable the debug assertion that enable extra checks on run time.

**Comment by varunagrawal** [comment]
Can you please share the changes here? I can run them locally.

**Comment by talregev** [comment]
I will. It will be later.

**Comment by varunagrawal** [comment]
> I will. It will be later.

Then can you please approve this PR?

**Review by talregev — APPROVED** [review]

**Pull request merged**
